Add blog post for v4.25 release (#4892)
tenzir/tenzir: web/blog/tenzir-node-v4.25/index.md

---
title: "Tenzir Node v4.25: Sinks Galore!"
slug: tenzir-node-v4.25
authors: [lava]

> date: 2025-01-06

Reviewer: ```suggestion
date: 2025-01-08
```

tags: [release, node]
comments: true
---

# Tenzir Node v4.25: Sinks Galore!

Tenzir Node v4.25 adds new sinks for Snowflake, OpenSearch, and Elasticsearch,
allowing seamless data integration and output. It also introduces major
enhancements to TQL2, including new language features and operator
improvements.

![Tenzir Node v4.25](tenzir-node-v4.25.excalidraw.svg)

[github-release]: https://github.com/tenzir/tenzir/releases/tag/v4.25.0

<!-- truncate -->

## TQL2

We continue enhancing TQL2 to make it the default language for writing
pipelines. This release expands TQL2's capabilities with numerous improvements
and features.

To simplify TQL2 adoption, we introduced a TQL2-only mode. When you enable
this mode for a Tenzir Nodes, all pipelines on that node run in TQL2, and
the Explorer interface automatically selects TQL2 mode. On such nodes, you
can only use TQL1 through the `legacy` operator.

To enable it, set `TENZIR_TQL2=true` in your environment,
configure `tenzir.tql2: true` in your settings, or start the node with
`tenzir-node --tql2`.

:::warning Call to Action
We're going to enable TQL2-only mode by default in the upcoming

> Tenzir Node 5.0 release.

Reviewer: ```suggestion
Tenzir Node v5.0 release.
```

Please try to test it out and report any issues back to us, so we can
ensure a seamless upgrade experience.
:::

### From and To Operators

With the [`from`](/next/tql2/operators/from) and
[`to`](/next/tql2/operators/to) operators, we have ported one of the final major
functionalities from TQL1 to TQL2.

The [`from`](/next/tql2/operators/from) operator lets you onboard data from
most sources effortlessly. For example, instead of

> ```

Reviewer: ```suggestion
```tql
```

load_http "https://example.com/file.json.gz"
decompress_gzip
read_json
```

you can now write

> ```

Reviewer: ```suggestion
```tql
```

from "https://example.com/file.json.gz"
```

to automatically determine the load operator, compression, and format.
If you want to learn more about how the automatic detection works, check
out the [reference documentation](/next/tql2/operators/from).

Conversely, the [`to`](/next/tql2/operators/to) operator simplifies data
delivery to most destinations. For instance, instead of

> ```

Reviewer: ```suggestion
```tql
```

write_json
compress "gzip"
save_file "myfile.json.gz"
```

you can now write

> ```

Reviewer: ```suggestion
```tql
```

separate operators for each compression algorithm:

 - [`compress_gzip`](/next/tql2/operators/compress_gzip)
 - [`compress_bz2`](/next/tql2/operators/compress_bz2)
 - [`compress_brotli`](/next/tql2/operators/compress_brotli)
 - [`compress_lz4`](/next/tql2/operators/compress_lz4)
 - [`compress_zstd`](/next/tql2/operators/compress_zstd)

as well as their respective `decompress_*` versions.

This allows us to expose more algorithm-specific options within each operator.
For example, the [`compress_gzip`](/next/tql2/operators/compress_gzip) and
[`decompress_gzip`](/next/tql2/operators/decompress_gzip) operators now
offer additional options, such as `compress_gzip level=10, format="deflate"`.

## New Sinks: Snowflake, OpenSearch, and Elasticsearch

Tenzir Node v4.25 offers several new sinks for sending your data. The
[`to_snowflake`](/next/tql2/operators/to_snowflake) sink enables seamless
writing of data into [Snowflake](https://www.snowflake.com) databases, helping
users integrate Tenzir with one of the most popular cloud data platforms.

![Snowflake Sink](snowflake.excalidraw.svg)

To send data to snowflake, a pipeline like this can be used:

> ```

Reviewer: ```suggestion
```tql
```

from {foo: 42, bar: true}
to_snowflake \
  account_identifier="asldyuf-xgb47555",
  user_name="tenzir_user",
  password="password1234",
  database="MY_DB",
  schema="MY_SCHEMA",
  table="TENZIR"
```

The new [`to_opensearch`](/next/tql2/operators/to_opensearch)
sink operator allow direct data output to [OpenSearch](https://opensearch.org/)
and OpenSearch Bulk API compatible providers such as
[Elasticsearch](https://www.elastic.co/elasticsearch).

![Opensearch Sink](opensearch.excalidraw.svg)

The [`to_opensearch`](/next/tql2/operators/to_opensearch) also
integrates with the new [`from`](/next/tql2/operators/from) and
[`to`](/next/tql2/operators/to) operators introduced above. So
it is possible to write to OpenSearch or ElasticSearch instances
by using the `opensearch:` or `elasticsearch:` URL schemes.

> ```

Reviewer: ```suggestion
```tql
```

---
title: "Tenzir Node v4.25: Sinks Galore!"
slug: tenzir-node-v4.25
authors: [lava]
date: 2025-01-06
tags: [release, node]
comments: true
---

# Tenzir Node v4.25: Sinks Galore!

Tenzir Node v4.25 adds new sinks for Snowflake, OpenSearch, and Elasticsearch, allowing seamless data integration and output. It also introduces major enhancements to TQL2, including new language features and operator improvements.

![Tenzir Node v4.25](tenzir-node-v4.25.excalidraw.svg)

[github-release]: https://github.com/tenzir/tenzir/releases/tag/v4.25.0

<!-- truncate -->

## New Sinks: Snowflake, OpenSearch, and Elasticsearch

Tenzir Node v4.25 offers several new sinks for sending your data. The [`to_snowflake`](/next/tql2/operators/to_snowflake) sink enables seamless writing of data into

> [Snowflake](https://www.snowflake.com) databases, helping users integrate Tenzir with one of the most popular cloud data platforms. Additionally, the
> new [`to_opensearch`](/next/tql2/operators/to_opensearch) and [`to_elasticsearch`](/next/tql2/operators/to_elasticsearch) sink operators allow direct data output to
> [OpenSearch](https://opensearch.org/) and [Elasticsearch](https://www.elastic.co/elasticsearch), two powerful search and analytics engines.

Reviewer 2: ```suggestion
[Snowflake](https://www.snowflake.com) databases, helping users integrate Tenzir with one of the most popular cloud data platforms. Additionally, the
new [`to_opensearch`](/next/tql2/operators/to_opensearch) sink operator allows direct data output to
[OpenSearch](https://opensearch.org/) and OpenSearch Bulk API compatible providers such as [Elasticsearch](https://www.elastic.co/elasticsearch), two powerful search and analytics engines.
```

For more information on using Tenzir in combination with these new sinks, check out our integration pages for [OpenSearch](/next/integrations/opensearch),

> [ElasticSearch](/next/integrations/elasticsearch) and [Snowflake](/next/integrations/snowflake)

Reviewer 2: Do we normally link to the `next` version of docs?

Reviewer 3: Yes, because otherwise the Docusaurus link check will fail. It's a framework deficiency.

Author: We have to for pages that are only introduced within the same release, otherwise the CI will fail because there are no versioned links for these yet.

### From and To

With the [`from`](/next/tql2/operators/from) and [`to`](/next/tql2/operators/to) operators, we have ported one of the last major functionalities from TQL1 to TQL2.

The [`from`](/next/tql2/operators/from) operator lets you onboard data from most sources effortlessly. For example, you can write `from "https://example.com/file.json.gz"` to automatically determine the load operator, compression, and format.

Conversely, the [`to`](/next/tql2/operators/to) operator makes it easy to send data to most destinations. For instance, you can write `to "ftps://example.com/file.json.gz"` to automatically determine the save operator, compression, and format.

### IPs and Timestamps

We improved support for working with the native IP and timestamp types in TQL2.

The `in` operator now checks IP or subnet data for subnet membership. For example, to check whether an IP address belongs to a subnet, use an expression like `1.2.3.4 in 1.2.0.0/16`. Similarly, to verify whether a subnet includes another subnet, write `1.2.0.0/16 in 1.0.0.0/8`. A new function, [`subnet(string)`](/next/tql2/functions/subnet), lets you parse strings as subnets.

> To streamline time-related computations, we added the [`from_epoch(x:duration) -> time`](/next/tql2/functions/from_epoch) function (#4854), which converts durations to epoch times. Additionally, you can now convert strings to durations with the [`duration(string)`](/next/tql2/functions/duration) function.

Reviewer 2: I think we can strip this section out and maybe add some more fluff to the bigger ones above.

We split the previously combined compress and decompress functionality into separate operators for Brotli and Gzip, giving users more granular control over compression tasks. The [`compress_gzip`](/next/tql2/operators/compress_gzip) and [`decompress_gzip`](/next/tql2/operators/decompress_gzip) operators now offer additional options, such as `compress_gzip level=10, format="deflate"`.

Finally, we renamed the TQL2 NICs operator from [`nic`](/next/tql2/operators/nic) to [`nics`](/next/tql2/operators/nics) to align with its documentation and maintain consistency.

## Bugfixes

This release also includes several important bug fixes. Highlights include:

- Resolved a parsing error with operator parenthesis continuation to ensure proper evaluation of expressions like `where (x or y) and z`.
- Fixed issues with handling empty records in `write_parquet`, enhancing compatibility with Parquet’s limitations.

> - Prevented skipping required positional arguments in the argument parser to ensure robust script execution.

Author: I don't think we should merge it, fixing a broken expression parser isn't really a new feature to be celebrated.

I'd rather just remove the whole section, but then there's no obvious place to put the changelog link. For now I've just adjusted the *highlights* language.
